File: doc/results/experiment_31012014_1512Errortest_FixedHc_smaller_than_dtdt/code/Walk.cpp
#include "main_walk.h"
using namespace std;

bool debug_walk = false;

Walk::Walk(int dimension, double _dt)
{
	if(debug_walk){cout<<"Walk::Walk"<<endl;}
	d = dimension;
	steps = 100;		/*Should not be specified here!*/
	x0 = 0; x1 = 1;
	y0 = 0; y1 = 1;
	z0 = 0; z1 = 1;
	D = 1;
	dt = _dt/steps;
	factor = sqrt(2*D*dt);
	rng = new Random();
	drift = factor/steps;
	inhomogenous = false;
};

void Walk::SetInitialCondition(int **C, int M, int N){
	/*Takes the initial condition as an array C[m,n] of ints(?) 
	describing how many walkers in each entry. */
	if(debug_walk){cout<<"Walk::SetInitialCondition"<<endl;}
	nwalkers = 0;
	m = M;
	n = N;
	for(int i=0; i<m;i++){
		for(int j=0; j<n;j++){
			nwalkers += C[i][j];
		}
	}
	walkers.resize(nwalkers);
	for(int i=0; i<nwalkers;i++){
		walkers[i] = new double[d];
		for(int l=0; l<d;l++){
			walkers[i][l] = 0;
		}
	}
	x = new double[m];
	y = new double[n];
	dx = (x1-x0)/(m-1);
	dy = (n>1)?((y1-y0)/(n-1)):0;
	for(int k=0;k<m;k++){
		x[k] = k*dx;
	}
	for(int k=0;k<n;k++){
		y[k] = k*dy;
	}
	int counter = 0;

	for(int i=0; i<m;i++){
		for(int j=0; j<n;j++){
			for(int l=0; l<C[i][j]; l++){
				PutWalkers(i,j,counter);
				counter ++;
			}
		}
	}
	x0_ = x0 - (dx/2.0);
	x1_ = x1 + (dx/2.0);
	y0_ = y0 - (dx/2.0);
	y1_ = y1 + (dx/2.0);
}

void Walk::ResetInitialCondition(int **C){
	if(debug_walk){cout<<"Walk:ResetInitialCondition"<<endl;}
	int nwalkers_prev = nwalkers;
	nwalkers = 0;
	for(int i=0; i<m;i++){
		for(int j=0; j<n;j++){
			nwalkers += C[i][j];
		}
	}
	if (nwalkers>nwalkers_prev){
		/*Do something smart*/
		walkers.resize(nwalkers);
	}
	// ResetWalkers();
	// cout<<"nwalkers = "<<nwalkers<<endl;
	for(int i=nwalkers_prev;i<nwalkers;i++)
		walkers[i] = new double[d];

	int counter = 0;
	for(int i=0; i<m;i++){
		for(int j=0; j<n;j++){
			for(int l=0; l<C[i][j]; l++){
				PutWalkers(i,j,counter);
				counter ++;
			}
		}
	}
}
bool Walk::HasLeftArea(double *pos){
	// if(debug_walk){cout<<"Walk::HasLeftArea"<<endl;}
	/*pos = [x] in 1d;
	pos = [x,y] in 2d etc*/
	if(d==1){
		if(pos[0]<x0 || pos[0]>x1){
			return true;
		}
	}
	else if(d==2){
		if(pos[0]>x1 || pos[0]<x0){
			return true;
		}
		if(pos[1]<y0 || pos[1]>y1){
			return true;
		}
	}
	return false;
}

void Walk::InhomogenousAdvance(int **C, double _dt){
	/*This should now implement the normal advance-function as well*/
	if(debug_walk){cout<<"Walk::InhomogenousAdvance"<<endl;}
	dt = _dt/steps;

	for(int i=0;i<m; i++){
		//Empty the array to conserve energy
		for(int j=0; j<n; j++){
			C[i][j] = 0;
		}
	}
	double DX[d];
	DX[0] = dx;
	if(d==2){DX[1] = dx;}
	// #pragma omp parallel 
	// {
	double *newPos, **s;
	newPos = new double[d];
	int *index = new int[d];
	double L = 0;
	double L_deriv = 0;
	double L0 = sqrt(2*dt);
	double L_deriv0 = (dt/d)/(2*sqrt(2*(dt/d)));
	double Tr,Tl,Delta_m,Delta_p,r,stepvector[d];
	if(not inhomogenous){
		L = L0*sqrt(D);
		Tr = 0.5;
		Delta_p = L;
		Delta_m = L;
	}
	int p=0;
	// #pragma omp for
	for(int i=0; i<nwalkers; i++){
		/*For every walker: */
		for(int j=0;j<steps;j++){
			/*Advance n steps*/
			FindPosition(walkers[i],index);
			p = int(round(rng->uniform()*(d-1)));	/*pick a spatial dimension to advance the walker*/
			// for(p=0; p<d; p++){
				if(inhomogenous){
					L = (d>1)?(L0*sqrt(aD[index[0]][index[1]])):(L0*sqrt(aD[index[0]][0]));
					/*This might work and is slightly better*/
					L_deriv = (d>1)?(L_deriv0/(DX[p]*sqrt(aD[index[0]][index[1]]))*(aDx[p][index[0]][index[1]])):
					(L_deriv0/(DX[p]*sqrt(aD[index[0]][0])))*(aDx[p][index[0]][0]);
					Tr = (1+0.5*L_deriv);
					Tl = (1-0.5*L_deriv);
					Delta_p = L*Tr;
					Delta_m = L*Tl;
					Tr /= 2.0;
					// cout<<"Delta_p,Delta_m = "<<Delta_p<<","<<Delta_m<<endl;
				}
				// r = ran0(&Idum);
				r = rng->uniform();
				if(r>Tr){
					stepvector[p] = Delta_p;
				}
				else{
					stepvector[p] = -Delta_m;
				}
			// }
			newPos = InhomogenousStep(walkers[i],stepvector);
			walkers[i] = checkpos(newPos,stepvector);
			stepvector[p] = 0;
		}
		FindPosition(walkers[i],index);
		if(d==1){
			C[index[0]][0] += 1;
		}
		else if(d==2){
			C[index[0]][index[1]] += 1;
		}
	}
// }
	// delete [] newPos,index;
}


double *Walk::InhomogenousStep(double *r, double *s){
	for(int l=0;l<d;l++){
		r[l] += s[l];
	}
	return r;
}

void Walk::SetDiffusionTensor(double **Diff, int M, int N){
	if(debug_walk){cout<<"Walk::SetDiffusionTensor"<<endl;}
	aD = new double*[M];
	dx = (x1-x0)/(M-1);
	dy = (N>1)?((y1-y0)/(N-1)):0;
	aDx.resize(d);
	for(int p=0;p<d;p++){
		aDx[p] = new double*[M];
		for(int i=0;i<M;i++){
			aDx[p][i] = new double[N];
		}
	}
	for(int i=0;i<M;i++){
		aD[i] = new double[N];
	}
	for(int i=0;i<M;i++){
		for(int j=0;j<N;j++){
			aD[i][j] = Diff[i][j];
		}
	}
	if(d==1){
		for(int i=1;i<M-1;i++){
			for(int j=0;j<1;j++){
				aDx[0][i][j] = (1.0/(2*dx))*(aD[i+1][j]-aD[i-1][j]);
			}
		}
		if(M!=0){
			aDx[0][0][0] = (1.0/dx)*(aD[1][0]-aD[0][0]);
			aDx[0][M-1][0] = (1.0/dx)*(aD[M-1][0]-aD[M-2][0]);
		}
	}
	if(d==2){
		for(int i=1;i<M-1;i++){
			for(int j=1;j<N-1;j++){
				aDx[0][i][j] = (1.0/(2*dx))*(aD[i+1][j]-aD[i-1][j]);
				aDx[1][i][j] = (1.0/(2*dy))*(aD[i][j+1]-aD[i][j-1]);
			}
		}
		for(int i=0;i<N;i++){
			aDx[0][0][i] = (1.0/dx)*(aD[1][i]-aD[0][i]);
			aDx[0][M-1][i] = (1.0/dx)*(aD[M-1][i]-aD[M-2][i]);
			if(i>0 && i<N-1){
				aDx[1][0][i] = (1.0/(2*dy))*(aD[0][i+1]-aD[0][i-1]);
				aDx[1][M-1][i] = (1.0/(2*dy))*(aD[M-1][i+1]-aD[M-1][i-1]);
			}
		}
		for(int i=0;i<M;i++){
			aDx[1][i][0] = (1.0/dy)*(aD[i][1]-aD[i][0]);
			aDx[1][i][N-1] = (1.0/dy)*(aD[i][N-2]-aD[i][N-1]);
			if(i>0 && i<M-1){
				aDx[0][i][0] = (1.0/(2*dx))*(aD[i+1][0]-aD[i-1][0]);
				aDx[0][i][N-1] = (1.0/(2*dx))*(aD[i+1][N-1]-aD[i-1][N-1]);
			}
		}
	}
	inhomogenous = true;
}

void Walk::SetDiffusionConstant(double Diff){
	if(debug_walk){cout<<"Walk::SetDiffusionConstant"<<endl;}
	D = Diff;
	inhomogenous = false;
}

void Walk::PutWalkers(int i, int j, int counter){
	// if(debug_walk){cout<<"Walk::PutWalkers"<<endl;}
	// walkers[counter][0] = x[i]+dx*(0.5-ran0(&Idum));
	walkers[counter][0] = x[i]+dx*(0.5-rng->uniform());
	if(d==2){
		// walkers[counter][1] = y[j]+dy*(0.5-ran0(&Idum));
		walkers[counter][1] = y[j]+dy*(0.5-rng->uniform());
	}
}


void Walk::FindPosition(double *pos, int *indx){
	/*Maps the walkers position to its index*/
	indx[0] = int(round(pos[0]/dx));
	if(d==2){
		indx[1] = int(round(pos[1]/dy));
	}
}

double *Walk::checkpos(double *r,double *s){
	/*Implements reflecting boundaries -- Need to adjust the boundaries by dx/2 so each thing has 
	as much space*/
	// tmp = r+s 	/*This is r in this case*/
	if(not HasLeftArea(r)){
		return r;
	}
	else{
		if(d ==1){
			if (r[0]<x0_){
				r[0] = x0_ - (r[0]-x0_);
				}
			else if(r[0]>x1_){
				r[0] = x1_ - (r[0]-x1_);
				}
			}
		else if(d == 2){
			if(r[0]<x0_){
				r[0] = x0_ - (r[0]-x0_);
				}
			else if( r[0]>x1_){
				r[0] = x1_ - (r[0]-x1_);
				}
			if (r[1]<y0_){
				r[1] = y0_ - (r[1]-y0_);
				}
			else if (r[1]>y1_){
				r[1] = y1_ - (r[1]-y1_);
				}
			}
		return r;
	}
}

File: doc/backup/walk.py
# -*- coding: utf-8 
# simple random walk program

import numpy as np


class Walk:
	"""
	Random walk class at the moment in 1D and 2D only
	"""
	def __init__(self,area,n=3,thresh=1.0,M=50):
		self.Hc = thresh			#conversion parameter between walkers and concentration
		self.d = 1 if len(area)==1 else 2 	#dimension of the walk
		self.area = area	#the coordinates of the area. [[x0,x1]] in 1D, [[x0,y0],[x1,y1]] in 2D
		self.M = M 					#conversion parameter
		self.walkers = []
		self.nwalkers = len(self.walkers)
		self.first = True
		self.factor = 1e-2			#reduces the steplength 
		if len(area)==1:
			self.x0,self.x1 = area[0]
		elif len(area) == 2:
			self.x0,self.y0 = area[0]
			self.x1,self.y1 = area[1]
		# print self.X


	def HasLeftArea(self,pos):
		"""pos = [x] in 1D; 
		pos = [x,y] in 2D"""
		# print 'HasLeftArea(self,pos)'
		if self.d==1:
			if pos[0]<self.x0 or pos[0]>self.x1:
				return True
		elif self.d==2:
			if pos[0]>self.x1 or pos[0]<self.x0:
				return True
			if pos[1]>self.y1 or pos[1]<self.y0:
				return True
		return False


	def advance(self,concentration):
		"""
		Concentration is now a matrix containing the entire area of the
		walk. This makes the coding simpler. The conversion should possibly 
		be done in the combine-class seeing as most random walk solvers do 
		not have this form of conversion...
		"""
		if self.d !=1:
			nx,ny = np.shape(concentration)
			self.X,self.Y = np.meshgrid(np.linspace(self.x0,self.x1,nx),\
			np.linspace(self.y0,self.y1,ny)) 	#should be only if needed
			self.dx = self.X[0,1]-self.X[0,0]
			self.dy = self.Y[1,0]-self.Y[0,0]
			self.x0_ = self.x0 - self.dx/2.0
			self.x1_ = self.x1 + self.dx/2.0
			self.y0_ = self.y0 - self.dy/2.0
			self.y1_ = self.y1 + self.dy/2.0
		else:
			self.x = np.linspace(self.x0,self.x1,len(concentration))
			self.dx = self.x[1]-self.x[0]
			self.dy = 0
			self.x0_ = self.x0 - self.dx/2.0
			self.x1_ = self.x1 + self.dx/2.0
		# print concentration
		a = self.InitializeTimestep(concentration)
		steps = 100
		indices = []
		walkers_leaving_area = []
		counter = 0
		if not a:
			return np.zeros(np.shape(concentration))
		for walker in xrange(self.nwalkers):
			"loop over all walkers"	
			# s = self.factor*np.random.standard_normal([self.d,steps])
			s = self.factor*(1.0 - 2.0*np.random.rand(self.d,steps)) 	#uniform distribution
			r0 = self.walkers[walker]

			for i in xrange(steps):
				r0[:] = self.checkpos(r0,s[:,i])
			self.walkers[walker][:] = r0[:]
			# if self.HasLeftArea(r0):
			# 	# print 'am I doing anything at all?'
			# 	indices.append(counter)
			# 	walkers_leaving_area.append(self.walkers[walker]) 
			# counter += 1
		boundary = self.ReturnBoundary(self.walkers,concentration)
		self.walkers = []
		self.nwalkers = len(self.walkers)
		boundary /= (self.M*self.Hc)
		return boundary + concentration

	
	def InitializeTimestep(self,C):
		"""C = concentration
		should work in 1d as well now"""
		nwalkers_tmp = self.nwalkers
		if nwalkers_tmp != 0: print 'this should not happen'
		C_tot = np.sum(C)
		N_walkers_tot = int(self.M*self.Hc*C_tot)
		# print 'N_walkers_tot = ',N_walkers_tot
		
		#There is some sort of leftover energy which must be accoounted for
		# print '"energy" left: %.16f'%(C_tot - N_walkers_tot/(self.M*self.Hc))

		if C_tot == 0 or N_walkers_tot <= 0:
			return None
		self.gradient = self.CalculateGradient(C)
		# Most walkers are lost here!!!

		if self.d ==1:
			for i in xrange(len(C)):
				n = int((C[i]/C_tot)*N_walkers_tot)
				C[i] -= n/(self.M*self.Hc)
				self.put_walkers(n,i,0)
		if self.d ==2:
			for i in xrange(len(C)):
				for j in xrange(len(C[i])):
					n = int((C[i][j]/C_tot)*N_walkers_tot)
					C[i][j] -= n/(self.M*self.Hc)
					self.put_walkers(n,i,j)
		self.nwalkers = len(self.walkers)
		# print 'self.walkers = %d, N_walkers_tot = %d'%(self.nwalkers,N_walkers_tot)
		lost = N_walkers_tot - self.nwalkers
		if lost >0:
			# print 'some %d walkers were lost, placing them'%lost
			while lost >0:
				if self.d==1:
					i = np.where(C==C.max())[0][0]
					C[i] -= 1/(self.M*self.Hc)
					j = 0
					# print 'i=',i, C[i], self.x[i]
				elif self.d==2:
					i,j = np.where(C==C.max())
					i = i[0]; j = j[0]
					C[i,j] -= 1/(self.M*self.Hc)
				lost -= 1
				self.put_walkers(1,i,j)
		return 1


	def put_walkers(self,N,i,j):
		"""Should work in 1d as well now"""
		if self.d==1:
			x = self.x[i]
			for k in xrange(N):
				self.walkers.append([x+self.factor*(0.5-np.random.uniform())])
		elif self.d==2:
			x = self.X[i,j]
			y = self.Y[i,j]
			for k in xrange(N):
				self.walkers.append([x+self.factor*(0.5-np.random.uniform()),\
					y+self.factor*(0.5-np.random.uniform())])

	def ReturnBoundary(self,walkers,concentration=None):
		"""Convert the number of walkers back into the concentration 
		at the boundary and return the boundary so it can be returned 
		from self.advance()"""
		# print concentration
		boundary = np.zeros(np.shape(concentration))
		# print 'starting %d walkers'%self.nwalkers
		if self.d ==1:
			dx = self.x[1]-self.x[0]
			# cont = 0
			for walker in walkers:
				# cont += 1
				index = self.FindPosition(walker,dx,0)
				boundary[index] += 1
		# print 'finished, have placed %d walkers'%np.sum(boundary)
		if self.d ==2:
			dx = self.X[0,1]-self.X[0,0]
			dy = self.Y[1,0]-self.Y[0,0]
			for walker in walkers:
				#find its position on the boundary
				index = self.FindPosition(walker,dx,dy)
				boundary[index[0],index[1]] += 1
		# print boundary, np.sum(boundary)
		return boundary

	def FindPosition(self,pos,dx,dy):
		"""Finds which index the walker belongs to. 
		Implements periodic boundary conditions on the walk-area
		Should work in 1d as well now"""
		indx = [-1,-1]
		if self.d==1:
			for i in xrange(len(self.x)+1):
				# print 'i = %d, x[i] = '%i,self.x[i],pos-self.x[i]
				if np.abs(pos-self.x[i])<dx/2.0:
					"This test must be implemented in 2D, and in self.checkpos()!"
					return i
		elif self.d==2:
			if pos[0]>self.x1_ or pos[0]<self.x0_:
				print 'this should not happen now! pos[x] = ',pos[0]
				pos[0] = np.fmod(pos[0],(self.X[0,-1]-self.X[0,0])+dx)+self.X[0,0]
				pos[0] *= ((self.X[0,-1]-self.X[0,0])+dx) if pos[0]<0 else 1
			for i in xrange(len(self.X)):
				if np.abs(pos[0]-self.X[i,i])<dx/2.0:
					indx[0] = i
					break
			if pos[1]>self.y1_ or pos[1]<self.y0_:
				print 'this should not happen now, pos[y] = ',pos[1]
				pos[1] = np.fmod(pos[1],(self.Y[-1,0]-self.Y[0,0])+dy)+self.Y[0,0]
				pos[1] *= ((self.Y[-1,0]-self.Y[0,0])+dx) if pos[1]<0 else 1
				# print pos[1]
			for j in xrange(len(self.Y)):
				if np.abs(pos[1]-self.Y[j,j])<dy/2.0:
					indx[1] = j
					break
			if indx[0] == -1 or indx[-1] == -1:
				print 'økadjs g'
			# print indx
			return indx

	def CalculateGradient(self,C):
		"""calculate the concentration gradient in an smart way
		Not so smart yet"""
		grad = np.zeros(np.shape(C))
		if self.d==1:
			for i in xrange(len(C)):
				grad[i] = (C[i]-C[i-1])/(self.x[1]-self.x[0])
		elif self.d==2:
			for i in xrange(len(C)):
				for j in xrange(len(C[i])):
					grad[i,j] = (C[i][j]-C[i-1][j])/((self.X[-1,-1]-self.X[0,0])/len(self.X))
					grad[i,j] = (C[i][j]-C[i][j-1])/((self.Y[-1,-1]-self.Y[0,0])/len(self.Y))
		return grad

	def checkpos(self,r,s):
		"""Implements reflecting boundaries -- Need to adjust the boundaries by dx/2 so each thing has 
		as much space"""
		tmp = r+s
		if not self.HasLeftArea(tmp):
			return tmp
		else:
			if self.d ==1:
				if tmp[0]<self.x0_:
					tmp[0] = self.x0_ - (tmp[0]-self.x0_)
				elif tmp[0]>self.x1_:
					tmp[0] = self.x1_ - (tmp[0]-self.x1_)
			elif self.d == 2:
				if tmp[0]<self.x0_:
					tmp[0] = self.x0_ - (tmp[0]-self.x0_)
				elif tmp[0]>self.x1_:
					tmp[0] = self.x1_ - (tmp[0]-self.x1_)
				if tmp[1]<self.y0_:
					tmp[1] = self.y0_ - (tmp[1]-self.y0_)
				elif tmp[1]>self.y1_:
					tmp[1] = self.y1_ - (tmp[1]-self.y1_)
			return tmp



import matplotlib.pyplot as mpl
import matplotlib.animation as animation
from mpl_toolkits.mplot3d import Axes3D 
# area = [[0.3,0.3],[0.4,0.4]]
t =0
def setup_plot():
	mpl.ion()
	fig  = mpl.figure()
	ax = fig.add_subplot(111,projection='3d')
	ax.set_autoscaley_on(False)
	return fig,ax

if __name__ == '__main__':
	if 1:
		"1D"
		n = 11
		U = np.zeros(n)
		U[:n/2] = 1
		x = np.linspace(0,1,n)
		area = [[0,1]]
		im = []
		fig  = mpl.figure()
		walk = Walk(area,1.0)
		while t<10:
			print 't = %d'%t, ' sum(U) = ',np.sum(U)
			im.append(mpl.plot(x,U,'b-'))
			U = walk.advance(U)	#[[1,1],[0,0]]
			t+=1
		ani = animation.ArtistAnimation(fig,im,interval=180,blit=True)
		mpl.show()
	if 0:
		"2D"
		X,Y = np.meshgrid(np.linspace(0,1,11),np.linspace(0,1,11))
		U = np.zeros((11,11))
		U[:11/2,:11/2] = 1
		area = [[0,0],[1,1]]
		fig,ax = setup_plot()
		walk = Walk(area,1.0)
		walk.factor = 7e-3
		wframe = ax.plot_wireframe(X,Y,U)
		mpl.draw()
		while t<1:
			print 't = %d'%t, ' sum(U) = ',np.sum(U)
			ax.collections.remove(wframe)
			U = walk.advance(U)	#[[1,1],[0,0]]
			t+=1
			wframe = ax.plot_wireframe(X,Y,U)
			mpl.draw()
